components/hab/src/command/pkg/upload.rs

```diff
@@ -52,26 +52,27 @@ use retry::retry;
 /// * Fails if it cannot find a package
 /// * Fails if the package doesn't have a `.hart` file in the cache
 /// * Fails if it cannot upload the file
-pub fn start<P: AsRef<Path>>(
+pub fn start<T, U>(
     ui: &mut UI,
     url: &str,
     channel: Option<&str>,
     token: &str,
-    archive_path: &P,
-    key_path: &P,
-) -> Result<()> {
+    archive_path: T,
+    key_path: U,
+) -> Result<()>
+where
+    T: AsRef<Path>,
+    U: AsRef<Path>,
```

Comment:
So the way this works is that the first time you match a generic it gets "bound" to the type which was passed to it. This means that if you pass an `&str` to the first instance of `P` and then a `Path` to the second, the function head won't get matched.

This function looks to be written before we knew how generics actually worked. You can see we re-used `P` and also referenced it with `&P` in the function signature even though we really just meant `P`